File: exercises/chapter1.cpp
#include <iostream>
#include <thread>
#include <RayTracerChallenge/datastructures/vec4.hpp>

class Projectile {
public:
  rtc::Vec4 pointPosition;
  rtc::Vec4 vecVelocity;

  Projectile(const rtc::Vec4 &pos, const rtc::Vec4 &vel) : pointPosition{pos},
    vecVelocity{vel} {
  }
};

class Environment {
public:
  rtc::Vec4 vecGravity;
  rtc::Vec4 vecWind;

  Environment(const rtc::Vec4 &grav, const rtc::Vec4 &wind) : vecGravity{grav},
    vecWind{wind} {
  }
};

void tick(const Environment &env, Projectile &proj) {
  const auto point = proj.pointPosition + proj.vecVelocity;
  const auto vel = proj.vecVelocity + env.vecGravity + env.vecWind;
  proj.pointPosition = point;
  proj.vecVelocity = vel;
}

int main() {
  std::cout << "Hello" << '\n';
  Projectile proj{rtc::point(0, 1, 0), rtc::vector(1, 1, 0).normalise()};
  const Environment env{rtc::vector(0, -0.1, 0), rtc::vector(-0.01, 0, 0)};

  while (proj.pointPosition.y > 0.000001) {
    std::this_thread::sleep_for(std::chrono::milliseconds(500));
    std::cout << proj.pointPosition << '\n';
    tick(env, proj);
  }
  std::cout << proj.pointPosition << '\n';
}

File: RTC.Console/Exercises/Chapter1.cs
using RTC.Console.Exercises.Structs;
using RTC.Datastructures;
using Environment = RTC.Console.Exercises.Structs.Environment;

namespace RTC.Console.Exercises;

public class Chapter1 : IChapterExercise
{
    public void Run()
    {
        var p = new Projectile
        {
            Position = Vec4.Point(0, 1, 0),
            Velocity = Vec4.Vector(1, 1, 0).Normalised
        };
        var e = new Environment
        {
            Gravity = Vec4.Vector(0, -0.1, 0),
            Wind = Vec4.Vector(-0.01, 0, 0)
        };

        while (p.Position.Y > 0)
        {
            System.Console.WriteLine(p.Position.ToString());
            p = Tick(e, p);
            Thread.Sleep(TimeSpan.FromSeconds(0.5));
        }
    }

    private static Projectile Tick(Environment env, Projectile proj)
    {
        return new Projectile
        {
            Position = proj.Position + proj.Velocity,
            Velocity = proj.Velocity + env.Gravity + env.Wind
        };
    }
}